### PR TITLE
Add drawer controller

### PR DESCRIPTION
**Pull request by isaiah658**
I made a few recent changes to the code such as utilizing an existing function in the drawers mod for inserting items in to a drawer and adding pipeworks support. Please look over the code any let me know if there are any issues or improvements. Thanks!

### CODE REVIEW

File: lua/controller.lua [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+]]--
+
+-- The gist of how the drawers mod stores data is that there are entities and the drawer node itself. The entities are needed to allow having multiple drawers in one node. The entities and node each store metadata about the item counts and such. It is necessary to change both at once otherwise in some cases the entity values are used and in other cases the node metadata is used.
+
+-- The gist of how the controller works is this. The drawer controller scans the adjacent tiles (length and height is configurable) and puts the item names and other info such as coordinates and the visualid of the entity in a table. That table is saved in the controllers metadata. The table is used to help prevent needing to scan all the drawers to deposit an item in certain situations. The table is only updated on an as needed basis, not by a specific time/interval. Controllers that have no items will not continue scanning drawers.
```
Comment (LNJ2) [inline lua/controller.lua]: Please only use about 80 chars per line. And for this I'd suggest you use a `--[[ ]]--` construct.

File: lua/controller.lua [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+
+local function controller_can_dig(pos, player)
+	local meta = minetest.get_meta(pos);
+	local inv = meta:get_inventory()
+	return inv:is_empty("src")
+end
+
+local function controller_allow_metadata_inventory_put(pos, listname, index, stack, player)
+	if minetest.is_protected(pos, player:get_player_name()) then
```
Comment (LNJ2) [inline lua/controller.lua]: Please run: `sed -i -e 's/minetest./core./' controller.lua`

File: LICENSE.txt [hunk 3]
```diff
@@ -108,6 +108,10 @@ Copyright (C) 2014 Justin Aquadro (MIT):
   textures/drawers_wood_front_2.png
   textures/drawers_wood_front_4.png
   textures/drawers_wood.png
+  
+Copyright (C) 2018 isaiah658 (CC0):
+  textures/drawer_controller_side.png
+  textures/drawer_controller_top_bottom.png
```
Comment (LNJ2) [inline LICENSE.txt]: Why don't you use the textures from the original mod? (It's ok of course to use yours :smiley:)

File: lua/controller.lua [hunk 4]
```diff
@@ -0,0 +1,377 @@
+--[[
+Minetest Mod Storage Drawers - A Mod adding storage drawers
+
+Copyright (C) 2018 isaiah658
+Copyright (C) 2018 LNJ <git@lnj.li>
```
Comment (LNJ2) [inline lua/controller.lua]: I actually haven't worked on this file in 2018, so please either use `2017-2018` or `2017`.

**Comment by isaiah658** [comment]
All the changes you mentioned were made. About the textures, I misread the readme description for the mod and didn't realize you created the original drawers mod for Minecraft! I thought you weren't associated with the Minecraft mod and were creating a clone of it for Minetest it. I really appreciate that you started working on making the drawers mod for Minetest and that you respond to issues and pull requests quickly. If myself or others find bugs in the mod, I will try to help fix them since you aren't interested in Minetest.

File: lua/controller.lua [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+core.register_node('drawers:controller', controller_def)
+
+-- Because the rest of the drawers mod doesn't have a hard depend on default, I changed the recipe to have an alternative
+if core.get_modpath("default") and default then
+	core.register_craft({
+		output = 'drawers:controller',
+		recipe = {
+			{'default:steel_ingot', 'default:steel_ingot', 'default:steel_ingot'},
+			{'default:tin_ingot', 'group:drawer', 'default:copper_ingot'},
+			{'default:steel_ingot', 'default:steel_ingot', 'default:steel_ingot'},
```
Comment (LNJ2) [inline lua/controller.lua]: I think the controller should cost at least one mese crystal or diamond. And please also add a recipe for MCL2.

**Comment by LNJ2** [comment]
Thanks for your work. I really appreciate that you write many comments (in code) and descriptive commit messages. :+1: :smiley:

**Comment by LNJ2** [comment]
One thing I noticed after merging: You don't use the translations, please fix that.

**Comment by isaiah658** [comment]
Can you explain more? The controller doesn't interfere or change any text on the drawers because it uses the same function that pipeworks would for inserting an item. Are you asking me to add parts of the controller to the template.pot file so others can add translations for it later on?

Also, if you request changes after a pull request is merged, do I open a new pull request or is this one still used?

**Comment by LNJ2** [comment]
Just do it as in here:
https://github.com/minetest-mods/drawers/blob/master/lua/api.lua#L30
https://github.com/minetest-mods/drawers/blob/master/lua/api.lua#L260

If you run `make poupdate` on the top level of the mod, all translation files will be updated.

You should create a new pull request and also reset your master branch and use a different branch for the new changes (using master isn't such a good idea, if you want to create more than one pull request).

**Pull request merged**